File: ad_library/src/main/java/com/adgrowth/adserver/RewardedAd.java
package com.adgrowth.adserver;

import android.app.Activity;
import android.content.DialogInterface;

import com.adgrowth.adserver.entities.RewardItem;
import com.adgrowth.internal.enums.AdType;
import com.adgrowth.internal.http.AdRequest;
import com.adgrowth.internal.interfaces.BaseAdListener;
import com.adgrowth.internal.views.BaseFullScreenAd;

public class RewardedAd extends BaseFullScreenAd<RewardedAd.Listener> {

    private static final int TIME_TO_REWARD = 30;
    private static final int TIME_TO_SHOW_TAP_TO_CLOSE = -3;

    private boolean mRewarded = false;

    public RewardedAd(String unitId) {
        this.mUnitId = unitId;
        this.mAdRequest = new AdRequest();
    }

    @Override
    public void onShow(DialogInterface dialogInterface) {
        super.onShow(dialogInterface);
        mDialog.showButtonText();
    }

    public void load(Activity context) {
        mRewarded = false;
        super.requestAd(context, AdType.REWARDED);
    }

    @Override
    protected void dismiss() {
        super.dismiss();
        mContext.runOnUiThread(() -> mListener.onDismissed());
    }

    @Override
    public void onDisplayTimeChanged(int currentTime) {
        super.onDisplayTimeChanged(currentTime);

        int remainingTime = (int) Math.ceil(TIME_TO_REWARD - currentTime);


        if (remainingTime < 0) {
            if (remainingTime <= TIME_TO_SHOW_TAP_TO_CLOSE) {
                mDialog.setButtonLabelText(mContext.getString(R.string.tap_to_close));
                stopAdStartedTimer();
            }
            return;
        }

        String buttonLabel = mContext.getResources().getQuantityString(R.plurals.remaining_seconds, remainingTime, remainingTime);

        if (remainingTime == 0) buttonLabel = mContext.getString(R.string.prize_received);

        mDialog.setButtonLabelText(buttonLabel);

        if ((currentTime >= TIME_TO_REWARD) && !mRewarded) {
            mRewarded = true;
            mListener.onEarnedReward(mAd.getReward());
        }

    }

    public interface Listener extends BaseAdListener<RewardedAd> {
        void onEarnedReward(RewardItem rewardItem);

        default void onDismissed() {
        }

    }
}

File: ad_library/src/main/kotlin/com/adgrowth/adserver/RewardedAd.kt
package com.adgrowth.adserver


import android.app.Activity
import com.adgrowth.adserver.entities.RewardItem
import com.adgrowth.adserver.exceptions.AdRequestException
import com.adgrowth.internal.integrations.RewardedManager
import com.adgrowth.internal.interfaces.integrations.RewardedIntegration
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class RewardedAd(unitId: String) : RewardedIntegration.Listener {
    private val mainScope = CoroutineScope(Dispatchers.Main)
    private lateinit var mContext: Activity
    private lateinit var mListener: Listener
    private var mAdManager: RewardedManager

    init {
        mAdManager = RewardedManager(unitId)
        mAdManager.listener = this
    }

    fun setListener(listener: Listener) {
        this.mListener = listener;
    }

    fun show(context: Activity) {
        mContext = context;
        if (!mAdManager.isLoaded) {
            mListener.onFailedToShow(AdRequestException.NOT_READY)
            return
        }
        mAdManager.show(context);
    }

    fun load(context: Activity) {
        mContext = context;
        mAdManager.load(context)
    }

    fun isLoaded(): Boolean {
        return mAdManager.isLoaded
    }

    fun isFailed(): Boolean {
        return mAdManager.isFailed
    }

    override fun onDismissed() {
        mainScope.launch { mListener.onDismissed() }
    }

    override fun onLoad(ad: RewardedIntegration) {
        mainScope.launch { mListener.onLoad(this@RewardedAd) }
    }

    override fun onFailedToLoad(exception: AdRequestException?) {
        mainScope.launch { mListener.onFailedToLoad(exception) }
    }

    override fun onClicked() {
        mainScope.launch { mListener.onClicked() }
    }

    override fun onFailedToShow(code: String?) {
        mainScope.launch { mListener.onFailedToShow(code) }
    }

    override fun onImpression() {
        mainScope.launch { mListener.onImpression() }
    }

    override fun onEarnedReward(rewardItem: RewardItem) {
        mainScope.launch { mListener.onEarnedReward(rewardItem) }
    }


    interface Listener {
        fun onLoad(ad: RewardedAd)
        fun onFailedToLoad(exception: AdRequestException?)
        fun onEarnedReward(rewardItem: RewardItem)

        @JvmDefault
        fun onDismissed() {
        }

        @JvmDefault
        fun onClicked() {
        }

        @JvmDefault
        fun onFailedToShow(code: String?) {
        }

        @JvmDefault
        fun onImpression() {
        }
    }
}
